File: src/rbtree.cpp
#include "rbtree.hpp"

#include <ac-library/containers/rbtree/persistent.hpp>
#include <cstdint>

namespace NACPyDiskStructs {
    using namespace NAC;

    PyObject* RBTreeOpen(PyObject*, PyObject* args) {
        const char* path;

        if (!PyArg_ParseTuple(args, "s", &path)) {
            return nullptr;
        }

        auto out = new TPersistentRBTree(path);
        out->FindRoot();
        return Py_BuildValue("K", (uintptr_t)out);
    }

    PyObject* RBTreeOpenRW(PyObject*, PyObject* args) {
        const char* path;

        if (!PyArg_ParseTuple(args, "s", &path)) {
            return nullptr;
        }

        auto out = new TPersistentRBTree(path, TFile::ACCESS_RDWR);
        out->FindRoot();
        return Py_BuildValue("K", (uintptr_t)out);
    }

    PyObject* RBTreeNew(PyObject*, PyObject* args) {
        const char* path;

        if (!PyArg_ParseTuple(args, "s", &path)) {
            return nullptr;
        }

        return Py_BuildValue("K", (uintptr_t)new TPersistentRBTree(path, TFile::ACCESS_CREATE));
    }

    PyObject* RBTreeDelete(PyObject*, PyObject* args) {
        unsigned long long int ptr;

        if (!PyArg_ParseTuple(args, "K", &ptr)) {
            return nullptr;
        }

        delete (TPersistentRBTree*)ptr;

        Py_RETURN_NONE;
    }

    PyObject* RBTreeInsert(PyObject*, PyObject* args) {
        unsigned long long int ptr;
        const char* key;
        Py_ssize_t keyLength;
        const char* value;
        Py_ssize_t valueLength;

        if (!PyArg_ParseTuple(args, "Ks#s#", &ptr, &key, &keyLength, &value, &valueLength)) {
            return nullptr;
        }

        ((TPersistentRBTree*)ptr)->Insert(
            TBlob(keyLength, key),
            TBlob(valueLength, value)
        );

        Py_RETURN_NONE;
    }

    PyObject* RBTreeGet(PyObject*, PyObject* args) {
        unsigned long long int ptr;
        const char* key;
        Py_ssize_t keyLength;

        if (!PyArg_ParseTuple(args, "Ks#", &ptr, &key, &keyLength)) {
            return nullptr;
        }

        if (auto value = ((TPersistentRBTree*)ptr)->Get(TBlob(keyLength, key))) {
            return Py_BuildValue("y#", value.Data(), value.Size());
        }

        Py_RETURN_NONE;
    }

    PyObject* RBTreeGetAll(PyObject*, PyObject* args) {
        unsigned long long int ptr;
        const char* prefix;
        Py_ssize_t prefixLength;

        if (!PyArg_ParseTuple(args, "Ks#", &ptr, &prefix, &prefixLength)) {
            return nullptr;
        }

        return Py_BuildValue("K", (uintptr_t)new TPersistentRBTree::TIterator(((TPersistentRBTree*)ptr)->GetAll(TBlob(prefixLength, prefix))));
    }

    PyObject* RBTreeIteratorDelete(PyObject*, PyObject* args) {
        unsigned long long int ptr;

        if (!PyArg_ParseTuple(args, "K", &ptr)) {
            return nullptr;
        }

        delete (TPersistentRBTree::TIterator*)ptr;

        Py_RETURN_NONE;
    }

    PyObject* RBTreeIteratorNext(PyObject*, PyObject* args) {
        unsigned long long int ptr;

        if (!PyArg_ParseTuple(args, "K", &ptr)) {
            return nullptr;
        }

        TBlob key;
        TBlob value;

        if (((TPersistentRBTree::TIterator*)ptr)->Next(key, value)) {
            auto* rv = PyTuple_New(2);

            PyTuple_SetItem(rv, 0, Py_BuildValue("y#", key.Data(), key.Size()));
            PyTuple_SetItem(rv, 1, Py_BuildValue("y#", value.Data(), value.Size()));

            return rv;
        }

        Py_RETURN_NONE;
    }
}

File: ac_diskstructs/rbtree.py
import ac_diskstructs.impl as LIB


class RBTreeIterator(object):
    def __init__(self, ptr):
        self.__ptr = ptr

    def delete(self):
        LIB.RBTreeIteratorDelete(self.__ptr)

        self.delete = lambda: None

    def __del__(self):
        self.delete()

    def __iter__(self):
        return self

    def __next__(self):
        item = LIB.RBTreeIteratorNext(self.__ptr)

        if item is None:
            self.delete()
            raise StopIteration()

        return item


def open_impl(path, rw, create):
    if create:
        return LIB.RBTreeNew(path)

    if rw:
        return LIB.RBTreeOpenRW(path)

    return LIB.RBTreeOpen(path)


class RBTree(object):
    def __init__(self, path, *, rw=False, create=False):
        self.__ptr = open_impl(path, rw, create)

    # Free memory
    def delete(self):
        LIB.RBTreeDelete(self.__ptr)

        self.delete = lambda: None

    def __del__(self):
        self.delete()

    def __getitem__(self, key):
        rv = LIB.RBTreeGet(self.__ptr, key)

        if rv is None:
            raise KeyError(key)

        return rv

    def get(self, key, default=None):
        try:
            return self[key]

        except KeyError:
            return default

    def __setitem__(self, key, value):
        LIB.RBTreeInsert(self.__ptr, key, value)

    def get_all(self, prefix):
        return RBTreeIterator(LIB.RBTreeGetAll(self.__ptr, prefix))
